cpp/src/arrow/io/file.cc

```diff
@@ -366,20 +383,20 @@ class OSFile {
 
  protected:
   Status SetFileName(const std::string& file_name) {
-    try {
 #if defined(_MSC_VER)
+    try {
       std::codecvt_utf8_utf16<wchar_t> utf16_converter;
       file_name_.assign(file_name, utf16_converter);
-#else
-      file_name_ = file_name;
-#endif
     } catch (boost::system::system_error& e) {
       return Status::Invalid(e.what());
     }
+#else
+      PlatformFilename::Init(file_name, &file_name_);
```

Comment:
I think it may be simpler to do `file_name_ = PlatformFilename{file_name}`, then you don't even need the `Init` method I don't think